core/src/main/java/com/linecorp/armeria/client/AbstractHttpRequestHandler.java

```diff
@@ -155,28 +157,33 @@ public final void operationComplete(ChannelFuture future) throws Exception {
         }
     }
 
-    final boolean tryInitialize() {
+    final boolean tryInitialize(boolean incrementNumUnfinishedResponses) {
         final HttpSession session = HttpSession.get(ch);
         id = session.incrementAndGetNumRequestsSent();
-        if (id >= MAX_NUM_REQUESTS_SENT || !session.canSendRequest()) {
-            final ClosedSessionException exception;
-            if (id >= MAX_NUM_REQUESTS_SENT) {
-                exception = new ClosedSessionException(
-                        "Can't send requests more than " + MAX_NUM_REQUESTS_SENT +
-                        " in one connection. ID: " + id);
-            } else {
-                exception = new ClosedSessionException(
-                        "Can't send requests. ID: " + id + ", session active: " +
-                        session.isAcquirable(responseDecoder.keepAliveHandler()));
+        if (incrementNumUnfinishedResponses) {
+            // Repeating the request after getting 417 expectation so just increment the number of requests.
+            assert session.incrementNumUnfinishedResponses();
```

Comment:
>  Is it safe to make this assertion?

Yes, it's safe. This is called when repeating and repeat is always called after removing the response which decreases the number by one.

Reply:
> This is called when repeating and repeat is always called after removing the response which decreases the number by one.

I see, I thought the max number of streams could be altered. Having said this, if you believe this scenario is unlikely I guess the current implementation is fine

> A SETTINGS frame MUST be sent by both endpoints at the start of a
   connection and MAY be sent at any other time by either endpoint over
   the lifetime of the connection.

ref: https://www.rfc-editor.org/rfc/rfc7540#section-6.5

Reply:
Ah ha, I got your point. 

So it's not true when
- The request is sent.
- The client receives settings frame which reduces the `maxConcurrentStreams` value.
- The client receives the response, reduces `maxUnfinishedResponses` by one, and tries to repeat.
- Even though `maxUnfinishedResponses` is reduced, the `maxConcurrentStreams` is altered so it might fail.

I couldn't think of this scenario and we definitely need to fix this. Thanks!